File: subjects/2023-2024-RO_VARIANT_I/alg_prog/CSharp/CSharp/communication/Communication.cs
namespace CSharp.encoder;

public class Communication
{
    private readonly Encoder code;
    private List<string> encodedMessages = [];

    public Communication(Encoder? e)
    {
        // NaiveEncoding as the default Encoder
        this.code = e ?? new NaiveEncoding();
    }

    public void addMessage(string m)
    {
        this.encodedMessages.Add(code.encode(m));
    }

    public List<string> getEncodedMessages()
    {
        return this.encodedMessages;
    }

    public override string ToString()
    {
        string result = "";

        for (var i = 0; i < this.encodedMessages.Count; i++)
        {
            result += this.encodedMessages[i];
            if (i != this.encodedMessages.Count - 1)
            {
                result += '#';
            }
        }

        return result;
    }
}

File: subjects/2023-2024-RO_VARIANT_I/alg_prog/Java/src/communication/Communication.java
package communication;

import encoder.Encoder;
import encoder.NaiveEncoding;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Communication {
    private final Encoder code;
    private final List<String> encodedMessages;

    public Communication(Encoder e) {
        // NaiveEncoding as default
        this.code = Objects.isNull(e) ? new NaiveEncoding() : e;
        this.encodedMessages = new ArrayList<>();
    }

    public void addMessage(String m) {
        this.encodedMessages.add(code.encode(m));
    }

    public List<String> getEncodedMessages() {
        return this.encodedMessages;
    }

    public String toString() {
        StringBuilder result = new StringBuilder();

        for (int i = 0; i < this.encodedMessages.size(); i++) {
            result.append(this.encodedMessages.get(i));
            if (i != this.encodedMessages.size() - 1) {
                result.append('#');
            }
        }

        return result.toString();
    }
}
